stdlib/public/core/Integers.swift.gyb

```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    if rhs < DoubleWidth((0, 0)) {
+      lhs <<= DoubleWidth((0, 0)) - rhs
+      return
+    }
+
+    // Shift is larger than this type's bit width.
+    if rhs._storage.high != 0 as High || rhs._storage.low >= DoubleWidth.bitWidth {
+      lhs = lhs < 0 ? ~0 : 0
+      return
+    }
+    
+    // Shift is exactly the width of `Base`, so high -> low.
+    if rhs._storage.low == numericCast(Base.bitWidth) {
+      lhs = DoubleWidth(
+        (lhs < 0 ? ~0 : 0, Low(extendingOrTruncating: lhs._storage.high)))
+      return
+    }
+
+    lhs &>>= rhs
+  }
+  
+  public static func &<<=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
+    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth - 1)
+
+    lhs._storage.high <<= High(rhs._storage.low)
+    if Base.bitWidth > rhs._storage.low {
+      lhs._storage.high |= High(extendingOrTruncating: lhs._storage.low >>
+        (numericCast(Base.bitWidth) - rhs._storage.low))
+    } else {
+      lhs._storage.high |= High(extendingOrTruncating: lhs._storage.low <<
+        (rhs._storage.low - numericCast(Base.bitWidth)))
+    }
+    lhs._storage.low <<= rhs._storage.low
+  }
+  
+  public static func &>>=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
+    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth - 1)
+
+    lhs._storage.low >>= rhs._storage.low
+    if Base.bitWidth > rhs._storage.low {
+      lhs._storage.low |= Low(extendingOrTruncating: lhs._storage.high <<
+        numericCast(numericCast(Base.bitWidth) - rhs._storage.low))
+    } else {
+      lhs._storage.low |= Low(extendingOrTruncating: lhs._storage.high >>
+        numericCast(rhs._storage.low - numericCast(Base.bitWidth)))
+    }
+    lhs._storage.high >>= High(extendingOrTruncating: rhs._storage.low)
+  }
+  
 % for x in chain(*binaryArithmetic.values()):
 
   // FIXME(integers): remove this once the operators are back to Numeric
   public static func ${x.operator} (
-    lhs: DoubleWidth<T>, rhs: DoubleWidth<T>
-  ) -> DoubleWidth<T> {
+    lhs: DoubleWidth, rhs: DoubleWidth
+  ) -> DoubleWidth {
     var lhs = lhs
     lhs ${x.operator}= rhs
     return lhs
   }
 
+%   argumentLabel = (x.firstArg + ':') if x.firstArg != '_' else ''
   public static func ${x.operator}=(
-    lhs: inout DoubleWidth<T>, rhs: DoubleWidth<T>
+    lhs: inout DoubleWidth, rhs: DoubleWidth
   ) {
-    fatalError()
+    let (result, overflow) = lhs.${x.name}ReportingOverflow(${argumentLabel}rhs)
+    _precondition(overflow == .none, "Overflow in ${x.operator}=")
+    lhs = result
   }
 % end
 
   public init(_truncatingBits bits: UInt) {
-    fatalError()
+    _storage.low = Low(_truncatingBits: bits)
+    _storage.high = High(_truncatingBits: bits >> UInt(Base.bitWidth))
   }
 
   // other
   //
   public init(_builtinIntegerLiteral x: _MaxBuiltinIntegerType) {
-    fatalError("Method must be overridden")
+    // FIXME: This won't work if `x` is out of range for `Int`
+    self.init(Int(_builtinIntegerLiteral: x))
```

Comment:
Will it change anything if we use `Int64` explicitly? I mean, it's not perfect still, but at least on 32-bit platforms it would support larger literals.